Collect fio data over time

noahl opened the pull request:
This PR adds a flag --run_for_minutes to the fio benchmark. When supplied, the user can specify a length of time (in minutes) to run for. Fio will then do repeated 10-minute jobs until it hits the user's requested time, reporting statistics for each job individually, and annotating each job's data with its job number and how many minutes into the overall run it started.

This PR also includes a bug fix in the case where fio is run against a formatted disk using a non-default device_fill_size parameter.

noahl commented:
@yuyantingzero @cmccoy Will you take a look?

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -280,6 +288,10 @@ def Prepare(benchmark_spec):
 
+  if FLAGS.run_for_minutes % MINUTES_PER_JOB != 0:
+    logging.warning('Runtime %s will be rounded up to the next multiple of %s '
+                    'minutes.' % (MINUTES_PER_JOB,))
```
Missing one argument?

noahl replied:
Yes, thanks for catching! Fixed.

cmccoy replied:
Use the logging module's built-in substitution here:
```{python}
logging.warning('Runtime %s will be ...', arg1, ...)
```

noahl replied:
Fixed to use built-in substitution. Thanks!

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -160,6 +165,9 @@ def GenerateFillCommand(fio_path, fill_path, fill_size):
 {% endfor %}
 """
 
+MINUTES_PER_JOB = 10  # Must match the JOB_FILE_TEMPLATE above.
```
Use this number in the job template, so we do not accidentally break the logic here.

noahl replied:
Done.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -340,21 +352,46 @@ def Run(benchmark_spec):
   fio_command = 'sudo %s --output-format=json %s' % (fio.FIO_PATH,
                                                      REMOTE_JOB_FILE_PATH)
+
+  run_reps = FLAGS.run_for_minutes // MINUTES_PER_JOB
```
This is not always correct. What if there are more than 1 scenarios in generated job file? Each scenario takes 10 minutes.

noahl replied:
The behavior I want is that each scenario is run for the given number of minutes, so I think this is actually the correct logic, but a confusing interface. I updated the help message for the --run_for_minutes option and added a comment here. Thanks for catching this!

yuyantingzero commented:
Could you add a test for fio benchmark itself on the repeated flow?

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -340,21 +352,46 @@ def Run(benchmark_spec):
                                                      REMOTE_JOB_FILE_PATH)
+
+  run_reps = FLAGS.run_for_minutes // MINUTES_PER_JOB
+  if FLAGS.run_for_minutes % MINUTES_PER_JOB != 0:
```
If users gives a customized job file and also gives FLAGS.run_for_minutes, the logic is broken and leads to unexpected behavior. The job file will be executed until run_reps repetitions. We can either make FLAGS.run_for_minutes more generic and work with customized job file, or we check fio_jobfile flag.

noahl replied:
Wow, this is bad. Thanks for catching it! I made the new logic run only for generated job files.

I thought about extending this flag to custom job files, but I don't see a good way to do it. We don't know how long a custom job file will last, so we can't figure out how many repeats we need. I could have made the flag number of repeats instead of minutes to run, but I think minutes are what we really care about - the number of minutes per job is just an implementation detail.

The only way I can see it working is if we start the custom job file and just keep running it until the total accumulated time passes the user's request, which might work. But I'm worried about measuring overhead somewhere in the system rather than actual fio run time, and fixing that would be very complex. I think that could be a separate project.

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -340,21 +352,46 @@ def Run(benchmark_spec):
+  samples = []
+  start_time = datetime.datetime.now()
+  for rep_num in xrange(run_reps):
+    logging.info('**** Repetition number %s of %s ****' % (rep_num, run_reps))
```
As above, no need for `%`

noahl replied:
Fixed. Thanks!

cmccoy commented on perfkitbenchmarker/packages/fio.py:
```diff
@@ -115,12 +115,13 @@ def FioParametersToJob(fio_parameters):
 
 
-def ParseResults(job_file, fio_json_result):
+def ParseResults(job_file, fio_json_result, base_metadata={}):
```
[Don't use mutable objects for argument defaults](http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments).

noahl replied:
Fixed.

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -298,13 +310,13 @@ def Prepare(benchmark_spec):
       fill_path = posixpath.join(mount_point, DEFAULT_TEMP_FILE_NAME)
-      if FLAGS['disk_fill_size'].present:
-        fill_size = FLAGS.disk_fill_size
+      if FLAGS['device_fill_size'].present:
```
Is this a bug fix?

noahl replied:
Yes, it is. There's another one in this PR too. I can move them to a separate PR if you want.

cmccoy replied:
Please do, or at least note the additional fixes in the PR description. Was the bug present in the last release?

noahl replied:
I don't think the bugs were in the last release. I'd like to keep the bugs in this PR, but I'll make a point of separating distinct changes in future PRs. Sorry for mixing the two this time.

cmccoy replied:
Could you update the description to note this fix?

noahl replied:
Updated. Sorry, I forgot to do this before.

noahl commented:
@yuyantingzero @cmccoy I added a test checking that the fio benchmark runs multiple times with the --run_for_minutes flag, and I believe I've addressed all of your other comments with the changes I just pushed.

voellm commented:
I plan to roll the next release today.  Is this ready to integrate?

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -90,11 +91,17 @@
+                     'minutes. Only valid when using --generate_scenarios. '
+                     'When using multiple scenarios, each one is run for the '
+                     'given number of minutes.',
+                     lower_bound=0)
```
Should this be 1? What happens when `fio` is set to run for 0 minutes?

noahl replied:
If you set it to run for 0 minutes, it should do a dry run. Everything looks exactly like a run, except that it never runs the fio command (and never tries to parse the results). I made this 0 because the behavior at 0 fits the pattern that happens at positive integers, so I didn't feel like I had a reason to forbid it. I can make it 1 - I don't have strong feelings about this.

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -90,11 +91,17 @@
+flags.DEFINE_integer('run_for_minutes', 10,
+                     'Repeat the job scenario(s) for the given number of '
+                     'minutes. Only valid when using --generate_scenarios. '
+                     'When using multiple scenarios, each one is run for the '
```
Could you document here the interaction between `--run_for_minutes` and `MINUTES_PER_JOB`?

noahl replied:
Done.

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -340,21 +355,59 @@ def Run(benchmark_spec):
+      stdout, stderr = vm.RemoteCommand(fio_command, should_log=True)
+
+      seconds_since_start = int(round((run_start - start_time).total_seconds()))
+      minutes_since_start = seconds_since_start // SECONDS_PER_MINUTE
```
Could you  `round(...)` here too?

noahl replied:
Done. I think the `round(...)` version is cleaner, but see what you think. I had to add `float(...)` calls to make it round properly.

cmccoy commented:
+1

noahl commented:
@yuyantingzero @cmccoy This branch is ready for review again.

cmccoy commented:
LGTM

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -90,11 +92,18 @@
+                     'minutes. Only valid when using --generate_scenarios. '
+                     'When using multiple scenarios, each one is run for the '
+                     'given number of minutes. Time will be rounded up to the '
+                     'next multiple of %s minutes.' % (MINUTES_PER_JOB,),
```
extra comma here.

noahl replied:
Removed the extra comma and the unnecessary parentheses.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -341,21 +356,58 @@ def Run(benchmark_spec):
+      run_start = datetime.datetime.now()
+      stdout, stderr = vm.RemoteCommand(fio_command, should_log=True)
+
+      seconds_since_start = int(round((run_start - start_time).total_seconds()))
```
Not necessary to create a new variable here if it's only used once.

noahl replied:
I'd like to leave the variable there just for readability. If I get rid of it, there will be a very large expression.

yuyantingzero commented:
2 small comments. LGTM
